sphinx_gherkindoc/glossary.py

```diff
@@ -45,34 +45,50 @@ def __eq__(self, other: object) -> bool:
 
 
 step_glossary: DefaultDict[str, GlossaryEntry] = defaultdict(GlossaryEntry)
+step_glossary_grouped: DefaultDict[str, DefaultDict[str, GlossaryEntry]] = defaultdict(
+    lambda: defaultdict(GlossaryEntry)
+)
 
 
-def make_steps_glossary(project_name: str) -> Optional[SphinxWriter]:
+def make_steps_glossary(
+    project_name: str, group_by: bool = False
+) -> Optional[SphinxWriter]:
     """Return SphinxWriter containing the step glossary information, if any."""
 
-    if not step_glossary:
+    if not step_glossary and not step_glossary_grouped:
         return None
 
     glossary = SphinxWriter()
     glossary.create_section(1, f"{project_name} Glossary")
 
-    master_step_names = {
-        name for gloss in step_glossary.values() for name in gloss.step_set
-    }
-    for term in sorted(master_step_names):
+    if group_by:
+        glossary.create_section(2, "Group By")
+
+        for step_type, glossary_group in step_glossary_grouped.items():
+            glossary.create_section(3, f"{step_type}")
+            _step_glossary(glossary, glossary_group)
+    else:
+        _step_glossary(glossary, step_glossary)
+
+    return glossary
```

Comment:
I had to stare at this for a while to work out the logic.
Conflicting interests:
1. Keeping the `if` in a postive form (`if x` is better than `if not x`)
2. Keeping the shorter code block as the `then` clause (not done here).

I was going to suggest preferring option 2 over option 1 as being net-simpler IMHO, as I was rewriting the code to be an example I kept stumbling over a deeper issue of how the code was refactored.
Boolean parameters are generally a code smell, two functions hiding together inside of one.
There are worse examples "in the wild", but it is bothering me.
HOWEVER, I am willing to say let's get this released and do a separate cleanup PR to refactor further since we now have some tests to make sure things still work.
I have some specific ideas about this, but I wanted to raise this so the author could also think about how to restructure it too.